projects/solar/test/test_sense_mcp3427.c

```diff
@@ -0,0 +1,327 @@
+    uint8_t addr = i % 8 + 1;
+    settings->mcp3427s[i].mcp3427_settings.addr_pin_0 = s_base_3_digits_to_pin[addr % 3];
+    settings->mcp3427s[i].mcp3427_settings.addr_pin_1 = s_base_3_digits_to_pin[addr / 3];
+  }
+  settings->num_mcp3427s = MAX_SOLAR_MCP3427;
+}
+
+// these is updated when |sense_register| is called
+static SenseCallback s_sense_callbacks[MAX_SOLAR_MCP3427];
+static void *s_sense_callback_contexts[MAX_SOLAR_MCP3427];
+static uint8_t s_num_sense_callbacks;
+
+StatusCode TEST_MOCK(sense_register)(SenseCallback callback, void *context) {
+  s_sense_callbacks[s_num_sense_callbacks] = callback;
+  s_sense_callback_contexts[s_num_sense_callbacks] = context;
+  s_num_sense_callbacks++;
+  return STATUS_CODE_OK;
+}
+
+// these are updated when the mcp3427 mocked functions are called
+static Mcp3427Callback s_mcp3427_callbacks[MAX_SOLAR_MCP3427];
+static void *s_mcp3427_callback_contexts[MAX_SOLAR_MCP3427];
+static Mcp3427FaultCallback s_mcp3427_fault_callbacks[MAX_SOLAR_MCP3427];
+static void *s_mcp3427_fault_callback_contexts[MAX_SOLAR_MCP3427];
+static uint8_t s_num_mcp3427_callbacks;
+static uint8_t s_times_mcp3427_start_called;
+
+// extract the callbacks and contexts from the MCP3427 storage
+StatusCode TEST_MOCK(mcp3427_start)(Mcp3427Storage *storage) {
+  s_mcp3427_callbacks[s_num_mcp3427_callbacks] = storage->callback;
+  s_mcp3427_callback_contexts[s_num_mcp3427_callbacks] = storage->context;
+  s_mcp3427_fault_callbacks[s_num_mcp3427_callbacks] = storage->fault_callback;
+  s_mcp3427_fault_callback_contexts[s_num_mcp3427_callbacks] = storage->fault_context;
+  s_num_mcp3427_callbacks++;
+  s_times_mcp3427_start_called++;
+  return STATUS_CODE_OK;
+}
+
+void setup_test(void) {
+  interrupt_init();
+  soft_timer_init();
+  event_queue_init();
+  data_store_init();
+  // The dependencies on I2C and sense are mocked out, so we don't initialize them
+
+  s_num_sense_callbacks = 0;
+  s_num_mcp3427_callbacks = 0;
+  s_times_mcp3427_start_called = 0;
+}
+void teardown_test(void) {}
+
+// Test that we can complete two normal sense cycles with one MCP3427.
+void test_sense_mcp3427_normal_cycle_one_mcp3427(void) {
+  bool is_set;
+  uint16_t set_value;
+  SenseMcp3427Settings settings = {
+    .mcp3427s = { {
+        .data_point = TEST_DATA_POINT,
+    } },
+    .num_mcp3427s = 1,
+  };
+  settings.mcp3427s[0].mcp3427_settings = s_test_mcp3427_settings;
+  TEST_ASSERT_OK(sense_mcp3427_init(&settings));
+  TEST_ASSERT_EQUAL(0, s_times_mcp3427_start_called);  // mcp3427 cycle not started yet
+
+  // start the MCP3427 cycles
+  TEST_ASSERT_OK(sense_mcp3427_start());
+  TEST_ASSERT_EQUAL(1, s_times_mcp3427_start_called);
+  data_store_get_is_set(TEST_DATA_POINT, &is_set);
+  TEST_ASSERT_EQUAL(false, is_set);  // not setting data yet
+
+  // make sure we received the various callbacks
+  TEST_ASSERT_NOT_NULL(s_sense_callbacks[0]);
+  TEST_ASSERT_NOT_NULL(s_mcp3427_callbacks[0]);
+  TEST_ASSERT_NOT_NULL(s_mcp3427_fault_callbacks[0]);
```

Comment:
IMO Move these checks into the mocked callback. You shouldn't be setting any nulls there I believe.